File: 백준/Gold/16398. 행성 연결/행성 연결.java
import java.io.*;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
    
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;
    static int[] parent;
    
    public static void main(String[] args) throws IOException {
        int n = Integer.parseInt(br.readLine());
        PriorityQueue<Edge> edges = new PriorityQueue<Edge>((e1, e2) -> e1.w - e2.w);
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                int w = Integer.parseInt(st.nextToken());
                if (i == j) {
                    continue;
                }
                edges.add(new Edge(i, j, w));
            }
        }
        parent = new int[n + 1];
        for (int i = 0; i < n + 1; i++) {
            parent[i] = i;
        }
        
        long result = 0;
        while (!edges.isEmpty()) {
            Edge edge = edges.poll();
            
            if (find(edge.v1) != find(edge.v2)) {
                union(edge.v1, edge.v2);
                result += edge.w;
            }
        }
        bw.write(result + "\n");
        bw.flush();
        bw.close();
    }
    
    static int find(int x) {
        if (x != parent[x]) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }
    
    static void union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        
        if (rootX < rootY) {
            parent[rootY] = rootX;
        } else if (rootX > rootY) {
            parent[rootX] = rootY;
        }
    }
    
    static class Edge {
        
        int v1;
        int v2;
        int w;
        
        public Edge(int v1, int v2, int w) {
            this.v1 = v1;
            this.v2 = v2;
            this.w = w;
        }
    }
}

File: 백준/Gold/16398. 행성 연결/행성 연결.py
import heapq
import sys

input = sys.stdin.readline


def find(x: int) -> int:
    if x != parent[x]:
        parent[x] = find(parent[x])
    return parent[x]


def union(x: int, y: int) -> None:
    root_x = find(x)
    root_y = find(y)

    if root_x < root_y:
        parent[root_y] = root_x
    elif root_x > root_y:
        parent[root_x] = root_y


n = int(input())
data = [list(map(int, input().split())) for i in range(n)]

parent = [i for i in range(n + 1)]
edges = []
for r in range(n):
    for c in range(r + 1, n):
        if r == c:
            continue
        heapq.heappush(edges, [data[r][c], r, c])

result = 0
while edges:
    edge = heapq.heappop(edges)
    w = edge[0]
    v1 = edge[1]
    v2 = edge[2]

    if find(v1) != find(v2):
        union(v1, v2)
        result += w
print(result)